### ISSUE TITLE
Create script to create pages/categories/etc needed for configuration

### ISSUE THREAD

**Issue by leoloso** [leoloso/PoP#38]
Most plug-ins need some configuration to be done, which involves creating pages (eg: POPTHEME_WASSUP_PAGE_ADDCONTENT) / categories (eg: POPTHEME_WASSUP_CAT_WEBPOSTS) / etc in the system, and then saving their IDs in a configuration file, such as environment-constants.php.

The creation of these is currently a manual process. It would be awesome to automate it, adding scripts to create the objects in the DB upon installation of the plug-in, and then automatically saving the IDs on the configuration file.

**Comment by fiskr** [comment]
I would like to do this, I like automating these sorts of things with scripts. :-)

**Comment by leoloso** [comment]
We will first tackle how to create pages and categories. These comprise around 95% of all IDs in the system. For other objects, such as default thumbnails, Google Map API Key, etc, we can refine it later on.

### What we gotta do

We need to re-engineer all `environment-constants.php` files so that, instead of hard-coding the IDs in, these are queried from the DB. 

So, we need to switch from:

`define ('POP_COREPROCESSORS_PAGE_MAIN', 19986);`

to:

`define ('POP_COREPROCESSORS_PAGE_MAIN', Call something to get the ID from the DB here);`

### How to do it

Creating all required DB objects can be done by hooking into `PoP:system-install`, which is executed each time the website is deployed: each plug-in,  in its file `installation.php`, can implement function `setup` hooking into that filter, and create all of its required pages/categories.

In function `setup`, it executes `PoP_Setup::create_page(...)` and `PoP_Setup::create_category(...)` for all needed pages and categories. Eg for plugin pop-coreprocessors:

```
// Namespace to avoid conflict with other plug-ins
$namespace = 'pop-coreprocessors';

// All pages developed in this plug-in
$all_pages = array('main', 'postauthors', 'recommendedcontent', ...);

// What pages are required by the website. Provide implementation data through hook `pop-coreprocessors:required-pages`, which returns an array of:
// key: "page-name" (eg: 'main')
// value: array of "page-data" (eg: array('title' => __('Main'), 'slug' => 'main', ...))
$required_pages = apply_filters('pop-coreprocessors:required-pages', array());

// Iterate all required_pages and, if they don't exist yet, create them
foreach ($required_pages as $page_name => $page_data) {
  if (in_array($page_name, $all_pages) && !PoP_Setup::page_exists($namespace, $page_name)) {
    PoP_Setup::create_page($namespace, $page_name, $page_data['title'], $page_data['slug'], $page_data['content'], $page_data['thumb']);
  }
}
```

Class `PoP_Setup` must also be implemented. Its function `create_page($namespace, $key, ...)` will create the page, given the information passed by parameters (title of the page, slug, content, thumb), and then save the new page ID in the DB. 

**Note about saving IDs in the DB:** all IDs for all pages will be saved in a unique array, so there is no need to do a DB query for each ID to fetch, under an entry from combining `$namespace` and `$key`. (Same for `create_category`)

To retrieve the ID on file environment-constants.php, we call function `PoP_Setup::get_page_id($namespace, $key)`. Eg: in pop-coreprocessors, we define: 

`define ('POP_COREPROCESSORS_PAGE_MAIN', PoP_Setup::get_page_id('pop-coreprocessors', 'main'));`

As mentioned, to create the required pages/categories, we must pass their information, such as:

- Page: title, slug, content, thumb
- Category: name, slug

This information must be passed by the implementation of the website. Eg: for the getpop-demo theme (provided in the source code), this implementation is done through plugin getpop-demo-environment. In that plugin, we create file `environment-setup.php`, calling all functions to create all pages/categories:

```
add_filter('pop-coreprocessors:required-pages', 'get_required_pages');
function get_required_pages($required_pages) {
  $required_pages['main'] = array('title' => __('Main'), 'slug' => 'main'));
  return $required_pages;
}
```

**Comment by leoloso** [comment]
Merged issue #49 (Improve automation through scripts) into this issue.

PoP currently does not have scripts for automating:

- The installation process (eg: download dependencies)
- Creation of needed pages/categories (addressed in #38)
- Compilation of Handlebars .tmpl files
- Mangling/minifying .css/.js files

Help from anyone with experience with the right tools (Gulp? Grunt? Composer? WP CLI?) will be appreciated!

**Comment by leoloso** [comment]
Merged issue #49 (Improve automation through scripts) into this issue.

There is some scripts set-up already, such as:

https://github.com/leoloso/PoP/blob/master/wp-content/themes/getpop-demo/build/minify.sh

However this has many problems, it is certainly not adequate:

- Environment variables must be declared in the system
- Dependency on installing [Google's minify](https://github.com/mrclay/minify) server in localhost

So something more proper is needed.

**Comment by leoloso** [comment]
No need for this anymore in the new PoP
